File: source/python/recommenders/matrixfactorization/normal_random.m


function rand_num = normal_random();

persistent spare;
persistent is_spare_ready;

% rand_num = rand();
% disp('rand_num');
% disp(rand_num);

if is_spare_ready
	is_spare_ready = false;
	rand_num = spare;
else
	u = rand() * 2 -1;
	v = rand() * 2 -1;
	s = u * u + v * v;

	while (s>= 1 || s==0)
		u = rand() * 2 -1;
		v = rand() * 2 -1;
		s = u * u + v * v;
	end

	mul = sqrt(-2 * log(s) /s);
	spare = v * mul;
	is_spare_ready = true;
	rand_num = u * mul;
end

File: source/python/utils/normal_random.py
import math

__author__ = 'fpena'

import numpy as np

class NormalRandom:

    spare = 0
    is_spare_ready = False

    @staticmethod
    def generate():

        # print('is_spare_ready', NormalRandom.is_spare_ready)
        # return np.random.random()

        if NormalRandom.is_spare_ready:
            # print('Inside if')
            NormalRandom.is_spare_ready = False
            return NormalRandom.spare
        else:
            # print('Inside else')
            u = np.random.random() * 2 - 1
            v = np.random.random() * 2 - 1
            s = u * u + v * v

            # print('u', u)
            # print('v', v)

            while s >= 1 or s == 0:
                # print('Inside while')
                u = np.random.random() * 2 - 1
                v = np.random.random() * 2 - 1
                s = u * u + v * v

                # print('u', u)
                # print('v', v)

            mul = math.sqrt(-2.0 * math.log(s) / s)
            NormalRandom.spare = v * mul
            NormalRandom.is_spare_ready = True

            return u * mul


    @staticmethod
    def generate_list(size):

        array = []
        for i in range(size):
            array.append(NormalRandom.generate())

        return np.array(array)


    @staticmethod
    def generate_matrix(rows, columns):

        matrix = []
        for i in range(rows):
            matrix.append(NormalRandom.generate_list(columns))

        # return np.array(matrix, dtype='float32')
        # return np.array(matrix, dtype=object)
        return np.array(matrix)

# np.random.seed(4)
#
# print np.random.rand()
# print NormalRandom.generate_list(10)
# my_matrix = NormalRandom.generate_matrix(3, 3)
# print(my_matrix)
# np.random.seed(4)
# temp_feature = np.random.rand(3, 3)
#
# print(temp_feature)
#
# print(np.array_equal(my_matrix, temp_feature))
# print(np.array_equal(temp_feature, my_matrix))


